hoss/dimension_utilities.py

```diff
@@ -75,8 +75,150 @@ def prefetch_dimension_variables(opendap_url: str, varinfo: VarInfoFromDmr,
+    """ Create an array containing the minimum and maximum bounds
+        for each pixel in a given dimension. 
+
+        The minimum and maximum values are determined under the assumption
+        that the dimension data is monotonically increasing and contiguous.
+        So for every bounds but the last, the bounds are simply extracted
+        from the dimension dataset.
+
+        The final bounds must be calculated with the assumption that 
+        the last data cell is edge-aligned and thus has a value the does
+        not account for the cell length. So, the final bound is determined
+        by taking the median of all the resolutions in the dataset to obtain
+        a resolution that can be added to the final data value. 
+
+        Ex: Input dataset with resolution of 3 degrees:  [ ... , 81, 84, 87]
+
+        Minimum | Maximum
+         <...>     <...>
+          81        84
+          84        87
+          87        ?  ->  87 + median resolution -> 87 + 3 -> 90
+        
+    """
+    # Access the dimension variable's data using the variable's full path.
+    dimension_array = prefetch_dataset[dimension_path][:]
+
+    median_resolution = np.median(np.diff(dimension_array))
+
+    # This array is the transpose of what is required, just for easier assignment
+    # of values (indices are [row, column]) during the bounds calculations:
+    cell_bounds = np.zeros(shape=(2, dimension_array.size), dtype=dimension_array.dtype)
+
+    # Minimum values are equal to the dimension pixel values (for lower left pixel alignment):
+    cell_bounds[0] = dimension_array[:]
+
+    # Maximum values are the next dimension pixel values (for lower left pixel alignment),
+    # so these values almost mirror the minimum values but start at the second pixel
+    # instead of the first. Here we calculate each bound except for the very last one.
+    cell_bounds[1][:-1] = dimension_array[1:]
+
+    # Last maximum value is the last pixel value (minimum) plus the median resolution:
+    cell_bounds[1][-1] = dimension_array[-1] + median_resolution
+
+    # Return transpose of array to get correct shape:
+    return cell_bounds.T
+
+
+def write_bounds(prefetch_dataset: Dataset,
+                 dimension_variable: VariableFromDmr) -> None:
+    """ Write the input bounds array to a given dimension dataset.
+        
+        First a new dimension is created for the new bounds variable
+        to allow the variable to be two-dimensional.
+        
+        Then the new bounds variable is created using two dimensions: 
+        (1) the existing dimension of the dimension dataset, and
+        (2) the new bounds variable dimension.
+    
+    """
+    # Create the bounds array.
+    bounds_array = get_bounds_array(prefetch_dataset, 
+                                    dimension_variable.full_name_path)
+
+    # Create the second bounds dimension.
+    dimension_group = '/' + '/'.join(dimension_variable.full_name_path.split('/')[1:-1])
```

Comment:
I wrestled with this one for a bit when I wrote it. The root group is always '/', so I expect the full name path to always begin with a slash. In ATL16, all the variables are in the root group so a full name path like `/npolar_grid_lat` would result in an empty string if a slash isn't added to the front.

Reply:
Ok, I might suggest looking at PurePosixPath from pathlib. (not only because we use it in datatree)

``` python
>>> from pathlib import PurePosixPath 
>>> test_paths = ['/', '/root', '/root/group1', '/root/group1/group2']
>>> dimension_name = [PurePosixPath(x).name for x in test_paths]
>>> dimension_path = [PurePosixPath(x).parent for x in test_paths]
>>> dimension_name
['', 'root', 'group1', 'group2']
>>> dimension_path
[PurePosixPath('/'), PurePosixPath('/'), PurePosixPath('/root'), PurePosixPath('/root/group1')]
```

This might be overkill, but I would understand right away what was being done,.

Reply:
I like this a lot better! Much more readable. I have to convert the output to a string since PurePosixPath returns a PurePosixPath object, but that's easy to do.

Reply:
I like it too, Owen was too psyched. :)

Reply:
edit, that should have said, "Owen was not too psyched". That kinda changes it a bit. LOL